python/ambassador/config/resourcefetcher.py

```diff
@@ -376,6 +367,154 @@ def handle_k8s_crd(self, obj: dict) -> None:
+        else:
+            ambassador_service_metadata: AnyDict = self.ambassador_service_raw.get('metadata', {})
+            ambassador_service_name = ambassador_service_metadata.get('name', None)
+            ambassador_service_namespace = ambassador_service_metadata.get('namespace', 'default')
+
+            if not ambassador_service_name:
+                self.logger.error(f"Unable to set Knative {kind} {name}'s load balancer, Ambassador service has no name to reference")
+            else:
+                # TODO: It is technically possible to use a domain other than
+                # cluster.local (common-ish on bare metal clusters). We can
+                # resolve the relevant domain by doing a DNS lookup on
+                # kubernetes.default.svc, but this problem appears elsewhere
+                # in the code as well and probably should just be fixed all at
+                # once.
+                current_lb_domain = f"{ambassador_service_name}.{ambassador_service_namespace}.svc.cluster.local"
+
+        observed_ingress: AnyDict = next(iter(status.get('privateLoadBalancer', {}).get('ingress', [])), {})
+        observed_lb_domain = observed_ingress.get('domainInternal')
+
+        has_new_lb_domain = current_lb_domain != observed_lb_domain
+
+        if has_new_generation or has_new_lb_domain:
+            utcnow = datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")
+            status = {
+                "observedGeneration": current_generation,
+                "conditions": [
+                    {
+                        "lastTransitionTime": utcnow,
+                        "status": "True",
+                        "type": "LoadBalancerReady"
+                    },
+                    {
+                        "lastTransitionTime": utcnow,
+                        "status": "True",
+                        "type": "NetworkConfigured"
+                    },
+                    {
+                        "lastTransitionTime": utcnow,
+                        "status": "True",
+                        "type": "Ready"
+                    }
+                ]
+            }
+
+            if current_lb_domain:
+                load_balancer = {
+                    "ingress": [
+                        {
+                            "domainInternal": current_lb_domain
+                        }
+                    ]
+                }
+
+                status['loadBalancer'] = load_balancer
+                status['privateLoadBalancer'] = load_balancer
+
+            status_update = (f"{kind.lower()}.networking.internal.knative.dev", namespace, status)
+
+            self.logger.info(f"Updating Knative {kind} {name} status to {status_update}")
+            self.aconf.k8s_status_updates[f"{name}.{namespace}"] = status_update
+        else:
+            self.logger.debug(f"Not reconciling Knative {kind} {name}: observed and current generations are in sync")
+
```

Comment:
This looks like it's mostly the old `cluster_ingresses_to_mapping` function? I'm curious why you inlined it, and what the changes between the two really are? (I can split it out and run `diff`, but that won't tell me what you were thinking. :wink:)

(Obviously one change is that `cluster_ingresses_to_mappings` wasn't in this file...)

Reply:
I'd be happy to move it to a separate function if it clarifies things. It was moved to this file so I could access the `status` of the resource initially, but then I also found the placement to be more sensible because treating external CRDs like IR (instead of emitting IR) didn't feel like the best fit.

Aside from that, the main changes are:
* We generate a Mapping per-host-and-service (what Knative calls a "split") combination (previously we were emitting one per split only, with the host as a regex).
* We handle the path and timeout from the Knative Ingress and pass it on to the Mapping (not previously handled at all).
* We used to write out an update to the status of Knative CRD for each split, which didn't make sense to me as it was updating the same key, but now it's just generated once.
* We set the load balancer endpoint if we can resolve it, same as with Kubernetes ingresses.
* The status update is only written if the generation or load balancer endpoint changes.

Reply:
Thanks for the summary! All of that sounds really good. Let's talk about structure on Slack a bit -- to me this does feel like multiple functions, but I want to make sure that I'm not just needlessly complicating the world. 🙂